# ENEM - Itens das provas de Matemática e Suas Tecnologias aplicadas entre os anos de 2010 a 2019

## Configuração do notebook

### Importação dos pacotes

In [1]:
import pandas as pd

In [2]:
import numpy as np

In [3]:
import json

### Criação do dicionário contendo os arquivos de cada ano

In [4]:
arquivos_itens_provas = {
    2010: '../2010/Dados/ITENS_PROVA_2010.csv',
    2011: '../2011/DADOS/ITENS_PROVA_2011.txt',
    2012: '../2012/DADOS/ITENS_ENEM_2012.csv',
    2013: {
        'ano': 2013,
        'caminho': '../2013/PLANILHAS/ITENS_ENEM_2013.xlsx',
        'colunas_posicao': ['Ordem_Amarelo', 'Ordem_Cinza', 'Ordem_Azul', 'Ordem_Rosa'],      
        'colunas_item': ['Cod.Amarelo', 'Cod.Cinza', 'Cod.Azul', 'Cod.Rosa'],
        'cores_provas': ['AMARELA', 'CINZA', 'AZUL', 'ROSA'],
        'codigos_provas': [179, 180, 181, 182]
    },
    2014: {
        'ano': 2014,
        'caminho': '../2014/PLANILHAS/ITENS_ENEM_2014.xlsx',
        'colunas_posicao': ['Ordem_Amarelo', 'Ordem_Cinza', 'Ordem_Azul', 'Ordem_Rosa'],      
        'colunas_item': ['Cod.Amarelo', 'Cod.Cinza', 'Cod.Azul', 'Cod.Rosa'],
        'cores_provas': ['AMARELA', 'CINZA', 'AZUL', 'ROSA'],
        'codigos_provas': [207, 208, 209, 210]
    },
    2015: {
        'ano': 2015,
        'caminho': '../2015/PLANILHAS/ITENS_ENEM_2015.xlsx',
        'colunas_posicao': ['Ordem_Amarelo', 'Ordem cinza', 'Ordem Azul', 'Ordem Rosa'],
        'colunas_item': ['Cod.Amarelo', 'Cod.Cinza', 'Cod.Azul', 'Cod.Rosa'],
        'cores_provas': ['AMARELA', 'CINZA', 'AZUL', 'ROSA'],
        'codigos_provas': [243, 244, 245, 246]
    },
    2016: '../2016/DADOS/itens_prova_2016.csv',
    2017: '../2017/DADOS/ITENS_PROVA_2017.csv',
    2018: '../2018/DADOS/ITENS_PROVA_2018.csv',
    2019: '../2019/DADOS/ITENS_PROVA_2019.csv',
}

Tendo em vista que a estrutura dos arquivos contendo esses dados mudou ao longo do tempo, vamos dividir a leitura dos arquivos em 3 etapas: de 2010 a 2012, de 2013 a 2015 e de 2016 a 2019.

### Criação da tabela final

Criação da lista com as colunas da tabela final

In [5]:
lista_colunas = [
    'ANO_PROVA',
    'CO_POSICAO',
    'SG_AREA',
    'CO_ITEM',
    'TX_GABARITO',
    'CO_HABILIDADE',
    'TX_COR',
    'CO_PROVA',
]

Criação da tabela final

In [6]:
itens_provas_mt_2010_2019 = pd.DataFrame(columns = lista_colunas)

Mostrando as informações da tabela final

In [7]:
itens_provas_mt_2010_2019.info()

<class 'pandas.core.frame.DataFrame'>
Index: 0 entries
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   ANO_PROVA      0 non-null      object
 1   CO_POSICAO     0 non-null      object
 2   SG_AREA        0 non-null      object
 3   CO_ITEM        0 non-null      object
 4   TX_GABARITO    0 non-null      object
 5   CO_HABILIDADE  0 non-null      object
 6   TX_COR         0 non-null      object
 7   CO_PROVA       0 non-null      object
dtypes: object(8)
memory usage: 0.0+ bytes


### Leitura dos arquivos de 2010 a 2012

Como os arquivos desses anos não possuem o mesmo padrão de conteúdo e nomes das colunas e nem de formato de arquivo, vamos realizar a leitura deles individualmente.

Realizando a leitura do arquivo do ano de 2010.

In [8]:
itens_2010 = pd.read_csv(arquivos_itens_provas[2010], sep=';')

Mostrando as informações do arquivo

In [9]:
itens_2010.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 740 entries, 0 to 739
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   CO_POSICAO     740 non-null    int64  
 1   SG_AREA        740 non-null    object 
 2   CO_ITEM        740 non-null    int64  
 3   TX_GABARITO    740 non-null    object 
 4   CO_HABILIDADE  740 non-null    int64  
 5   TX_COR         740 non-null    object 
 6   CO_PROVA       740 non-null    int64  
 7   TP_LINGUA      10 non-null     float64
dtypes: float64(1), int64(4), object(3)
memory usage: 46.4+ KB


Selecionando apenas as linhas dos itens da área de Matemática e Suas Tecnologias

In [10]:
itens_2010 = itens_2010[itens_2010['SG_AREA'] == 'MT']

Excluindo a coluna chamada "TP_LINGUA" que é específica para itens das provas de Língua Estrangeira

In [11]:
itens_2010 = itens_2010.drop('TP_LINGUA', axis=1)

Incluindo a coluna com o ano da prova e preenchendo com o valor 2010

In [12]:
itens_2010['ANO_PROVA'] = 2010

Mostrando as cinco primeiras linhas da tabela dos itens de 2010

In [13]:
itens_2010.head()

CO_POSICAO SG_AREA  CO_ITEM TX_GABARITO  CO_HABILIDADE   TX_COR  \
560         136      MT    70114           C              1  AMARELO   
561         137      MT    72709           E              7  AMARELO   
562         138      MT    70700           E             11  AMARELO   
563         139      MT    70379           B             12  AMARELO   
564         140      MT    72470           C             25  AMARELO   

     CO_PROVA  ANO_PROVA  
560        97       2010  
561        97       2010  
562        97       2010  
563        97       2010  
564        97       2010

Incluindo os registros do ano de 2010 na tabela final

In [14]:
itens_provas_mt_2010_2019 = itens_provas_mt_2010_2019.append(itens_2010)

Mostrando os cinco últimos registros da tabela para confirmar que os registros do ano de 2010 foram adicionados

In [15]:
itens_provas_mt_2010_2019.tail()

ANO_PROVA CO_POSICAO SG_AREA CO_ITEM TX_GABARITO CO_HABILIDADE TX_COR  \
735      2010        176      MT   72775           B            16   ROSA   
736      2010        177      MT   70845           E            18   ROSA   
737      2010        178      MT   70765           D             8   ROSA   
738      2010        179      MT   73400           D             2   ROSA   
739      2010        180      MT   59475           E            24   ROSA   

    CO_PROVA  
735      100  
736      100  
737      100  
738      100  
739      100

Realizando a leitura do arquivo do ano de 2011.

In [16]:
itens_2011 = pd.read_csv(arquivos_itens_provas[2011], header=None)

Mostrando as informações do arquivo

In [17]:
itens_2011.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 925 entries, 0 to 924
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       925 non-null    object
dtypes: object(1)
memory usage: 7.4+ KB


Como o arquivo veio em formato de texto com apenas uma coluna, vamos extrair as informações de cada coluna seleciuonando os caracteres da string da coluna

In [18]:
itens_2011['CO_POSICAO'] = itens_2011[0].str[:3]
itens_2011['SG_AREA'] = itens_2011[0].str[3:5]
itens_2011['CO_ITEM'] = itens_2011[0].str[5:10]
itens_2011['TX_GABARITO'] = itens_2011[0].str[10]
itens_2011['CO_HABILIDADE'] = itens_2011[0].str[11:13]
itens_2011['TX_COR'] = itens_2011[0].str[13:20]
itens_2011['CO_PROVA'] = itens_2011[0].str[20:]

Mostrando as informações do arquivo

In [19]:
itens_2011.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 925 entries, 0 to 924
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   0              925 non-null    object
 1   CO_POSICAO     925 non-null    object
 2   SG_AREA        925 non-null    object
 3   CO_ITEM        925 non-null    object
 4   TX_GABARITO    925 non-null    object
 5   CO_HABILIDADE  925 non-null    object
 6   TX_COR         925 non-null    object
 7   CO_PROVA       925 non-null    object
dtypes: object(8)
memory usage: 57.9+ KB


Agora vamos excluir a coluna com o texto original

In [20]:
itens_2011 = itens_2011.drop(0, axis=1)

Selecionando apenas as linhas dos itens da área de Matemática e Suas Tecnologias

In [21]:
itens_2011 = itens_2011[itens_2011['SG_AREA'] == 'MT']

Incluindo a coluna com o ano da prova e preenchendo com o valor 2011

In [22]:
itens_2011['ANO_PROVA'] = 2011

Mostrando os cinco últimos registros da tabela para confirmar que os registros do ano de 2011 foram adicionados

In [23]:
itens_2011.head()

CO_POSICAO SG_AREA CO_ITEM TX_GABARITO CO_HABILIDADE   TX_COR CO_PROVA  \
560        136      MT   70183           A             1  Cinza        130   
561        137      MT   72052           B            10  Cinza        130   
562        138      MT   72676           E            21  Cinza        130   
563        139      MT   71649           E             5  Cinza        130   
564        140      MT   72045           E            17  Cinza        130   

     ANO_PROVA  
560       2011  
561       2011  
562       2011  
563       2011  
564       2011

Incluindo os registros do ano de 2011 na tabela final

In [24]:
itens_provas_mt_2010_2019 = itens_provas_mt_2010_2019.append(itens_2011)

Mostrando os cinco últimos registros da tabela para confirmar que os registros do ano de 2012 foram adicionados

In [25]:
itens_provas_mt_2010_2019.tail()

ANO_PROVA CO_POSICAO SG_AREA CO_ITEM TX_GABARITO CO_HABILIDADE   TX_COR  \
920      2011        176      MT   72041           C             9  Cinza     
921      2011        177      MT   70186           C            25  Cinza     
922      2011        178      MT   72980           C             9  Cinza     
923      2011        179      MT   71221           A            12  Cinza     
924      2011        180      MT   71071           B            15  Cinza     

    CO_PROVA  
920      136  
921      136  
922      136  
923      136  
924      136

Realizando a leitura do arquivo do ano de 2012.

In [26]:
itens_2012 = pd.read_csv(arquivos_itens_provas[2012])

Mostrando as informações do arquivo

In [27]:
itens_2012.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 925 entries, 0 to 924
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   ID_POSICAO     925 non-null    object
 1   COD_AREA       925 non-null    object
 2   ID_ITEM        925 non-null    int64 
 3   DS_GABARITO    925 non-null    object
 4   ID_HABILIDADE  925 non-null    int64 
 5   DS_COR         925 non-null    object
 6   ID_PROVA       925 non-null    int64 
dtypes: int64(3), object(4)
memory usage: 50.7+ KB


Selecionando apenas as linhas dos itens da área de Matemática e Suas Tecnologias

In [28]:
itens_2012 = itens_2012[itens_2012['COD_AREA'] == 'MT']

Renomeando as colunas para manter o padrão

In [29]:
itens_2012 = itens_2012.rename(columns={
    'ID_POSICAO': 'CO_POSICAO', 
    'COD_AREA': 'SG_AREA', 
    'ID_ITEM': 'CO_ITEM',
    'DS_GABARITO': 'TX_GABARITO',
    'ID_HABILIDADE': 'CO_HABILIDADE',
    'DS_COR': 'TX_COR',
    'ID_PROVA': 'CO_PROVA'
})

Incluindo a coluna com o ano da prova e preenchendo com o valor 2012

In [30]:
itens_2012['ANO_PROVA'] = 2012

Mostrando as cinco primeiras linhas da tabela dos itens de 2010

In [31]:
itens_2012.head()

CO_POSICAO SG_AREA  CO_ITEM TX_GABARITO  CO_HABILIDADE   TX_COR  CO_PROVA  \
560          1      MT    10547           A              4  Amarelo       149   
561          2      MT    15402           D             11  Amarelo       149   
562          3      MT    15460           E             30  Amarelo       149   
563          4      MT     9909           D             12  Amarelo       149   
564          5      MT     8444           E             20  Amarelo       149   

     ANO_PROVA  
560       2012  
561       2012  
562       2012  
563       2012  
564       2012

Incluindo os registros do ano de 2012 na tabela final

In [32]:
itens_provas_mt_2010_2019 = itens_provas_mt_2010_2019.append(itens_2012)

Mostrando os cinco últimos registros da tabela para confirmar que os registros do ano de 2012 foram adicionados

In [33]:
itens_provas_mt_2010_2019.tail()

ANO_PROVA CO_POSICAO SG_AREA CO_ITEM TX_GABARITO CO_HABILIDADE  \
920      2012         41      MT    8405           D            25   
921      2012         42      MT   12730           C             1   
922      2012         43      MT   14393           A            25   
923      2012         44      MT    8428           B            26   
924      2012         45      MT    7282           A            16   

            TX_COR CO_PROVA  
920  Cinza (Ledor)      156  
921  Cinza (Ledor)      156  
922  Cinza (Ledor)      156  
923  Cinza (Ledor)      156  
924  Cinza (Ledor)      156

Mostrando os valores encontrados na coluna "ANO_PROVA"

In [34]:
itens_provas_mt_2010_2019['ANO_PROVA'].value_counts()

2011    225
2012    225
2010    180
Name: ANO_PROVA, dtype: int64

### Leitura dos arquivos de 2013 a 2015

Vamos criar uma função para ler automaticamente os arquivos de 2013 a 2015 e criar uma tabela única

In [35]:
def criar_novo_dataframe(arquivo: dict) -> pd.DataFrame:
    """Cria um novo dataframe a partir do dicionário do arquivo do ano"""
    novo_df = pd.read_excel(arquivo['caminho'], sheet_name='MTT')
    itens_provas = pd.DataFrame(columns = lista_colunas)
    itens_provas['CO_POSICAO'] = np.append(
        np.append(novo_df[arquivo['colunas_posicao'][0]], novo_df[arquivo['colunas_posicao'][1]]),
        np.append(novo_df[arquivo['colunas_posicao'][2]], novo_df[arquivo['colunas_posicao'][3]])
    )
    itens_provas['SG_AREA'] = 'MT'
    itens_provas['CO_ITEM'] = np.append(
        np.append(novo_df[arquivo['colunas_item'][0]], novo_df[arquivo['colunas_item'][1]]),
        np.append(novo_df[arquivo['colunas_item'][2]], novo_df[arquivo['colunas_item'][3]])
    )
    itens_provas['TX_GABARITO'] = np.append(
        np.append(novo_df['GAB'], novo_df['GAB.1']),
        np.append(novo_df['GAB.2'], novo_df['GAB.3'])
    )
    itens_provas['CO_HABILIDADE'] = np.append(
        np.append(novo_df['HAB'], novo_df['HAB.1']),
        np.append(novo_df['HAB.2'], novo_df['HAB.3'])
    )
    itens_provas['TX_COR'] = np.array([cor for cor in arquivo['cores_provas'] for i in range(novo_df.shape[0])])
    itens_provas['CO_PROVA'] = np.array([cor for cor in arquivo['codigos_provas'] for i in range(novo_df.shape[0])])
    itens_provas['ANO_PROVA'] = arquivo['ano']
    return itens_provas

Agora vamos fazer a leitura do arquivo e adicionar os registros das tabelas dos anos de 2013 a 2015 diretamente na tabela final

In [36]:
anos_provas_2013_2015 = [2013, 2014, 2015]
for ano in anos_provas_2013_2015:
    novo_df = criar_novo_dataframe(arquivos_itens_provas[ano])
    itens_provas_mt_2010_2019 = itens_provas_mt_2010_2019.append(novo_df)

Mostrando as informações da tabela final após a edição

In [37]:
itens_provas_mt_2010_2019.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1170 entries, 560 to 179
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   ANO_PROVA      1170 non-null   object
 1   CO_POSICAO     1170 non-null   object
 2   SG_AREA        1170 non-null   object
 3   CO_ITEM        1170 non-null   object
 4   TX_GABARITO    1170 non-null   object
 5   CO_HABILIDADE  1170 non-null   object
 6   TX_COR         1170 non-null   object
 7   CO_PROVA       1170 non-null   object
dtypes: object(8)
memory usage: 82.3+ KB


Mostrando os valores encontrados na coluna "ANO_PROVA"

In [38]:
itens_provas_mt_2010_2019['ANO_PROVA'].value_counts()

2011    225
2012    225
2010    180
2013    180
2014    180
2015    180
Name: ANO_PROVA, dtype: int64

### Leitura dos arquivos de 2016 a 2019

Como os arquivos com os dados dos anos de 2016 a 2016 possuem até mais colunas do que as desejadas, vamos fazer a leitura, excluir as colunas que não serão utilizadas e inserir os registros na tabela final

In [39]:
anos_provas_2016_2019 = [2016, 2017, 2018, 2019]
for ano in anos_provas_2016_2019:
    novo_df = pd.read_csv(arquivos_itens_provas[ano], sep=';')
    novo_df = novo_df[novo_df['SG_AREA'] == 'MT']
    novo_df['ANO_PROVA'] = ano
    novo_df = novo_df.drop('TP_LINGUA', axis=1)
    novo_df = novo_df.drop('IN_ITEM_ADAPTADO', axis=1)
    itens_provas_mt_2010_2019 = itens_provas_mt_2010_2019.append(novo_df)

Vamos renumerar os índices das linhas da tabela final

In [40]:
itens_provas_mt_2010_2019 = itens_provas_mt_2010_2019.reset_index(drop=True)

Mostrando as informações para comprovar que os índices agora vão de 0 a 3149, totalizando 3150 itens

In [41]:
itens_provas_mt_2010_2019.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3150 entries, 0 to 3149
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   ANO_PROVA      3150 non-null   object
 1   CO_POSICAO     3150 non-null   object
 2   SG_AREA        3150 non-null   object
 3   CO_ITEM        3150 non-null   object
 4   TX_GABARITO    3150 non-null   object
 5   CO_HABILIDADE  3150 non-null   object
 6   TX_COR         3150 non-null   object
 7   CO_PROVA       3150 non-null   object
dtypes: object(8)
memory usage: 197.0+ KB


In [42]:
itens_provas_mt_2010_2019

ANO_PROVA CO_POSICAO SG_AREA CO_ITEM TX_GABARITO CO_HABILIDADE   TX_COR  \
0         2010        136      MT   70114           C             1  AMARELO   
1         2010        137      MT   72709           E             7  AMARELO   
2         2010        138      MT   70700           E            11  AMARELO   
3         2010        139      MT   70379           B            12  AMARELO   
4         2010        140      MT   72470           C            25  AMARELO   
...        ...        ...     ...     ...         ...           ...      ...   
3145      2019        176      MT   63359           B             9     Rosa   
3146      2019        177      MT   15171           D            25     Rosa   
3147      2019        178      MT   97573           E            10     Rosa   
3148      2019        179      MT   25163           D            25     Rosa   
3149      2019        180      MT   81332           E             9     Rosa   

     CO_PROVA  
0          97  
1          97  
2          97  
3          97  
4          97  
...       ...  
3145      558  
3146      558  
3147      558  
3148      558  
3149      558  

[3150 rows x 8 columns]

### Verificando a integridade dos dados

Mostrando que todos os anos foram adicionados na tabela final

In [43]:
itens_provas_mt_2010_2019['ANO_PROVA'].value_counts(sort=False)

2010    180
2011    225
2012    225
2013    180
2014    180
2015    180
2016    630
2017    450
2018    450
2019    450
Name: ANO_PROVA, dtype: int64

Vamos verificar os valores dos códigos das posições

In [44]:
itens_provas_mt_2010_2019['CO_POSICAO'].describe()

count     3150
unique     180
top        136
freq        36
Name: CO_POSICAO, dtype: int64

Desta forma, das 3150 linhas da tabela final, apenas 732 são únicas, ou seja, apenas 732 itens de Matemática diferentes foram utilizados nas provas dos últimos 10 anos

Vamos conferir se todos os itens são realmente da área de Matemática e Suas Tecnologias

In [45]:
itens_provas_mt_2010_2019['SG_AREA'].unique()

array(['MT'], dtype=object)

Como os itens são repetidos e trocados de ordem em provas com as cores diferentes, vamos verificar a quantidade de itens únicos utilizando a coluna com o código do item ("CO_ITEM")

In [46]:
itens_provas_mt_2010_2019['CO_ITEM'].describe()

count      3150
unique      732
top       32010
freq          9
Name: CO_ITEM, dtype: int64

Vamos verificar se as letras presentes no texto do gabarito são apenas letras de A até E e a letra X para os itens anulados

In [47]:
itens_provas_mt_2010_2019['TX_GABARITO'].unique()

array(['C', 'E', 'B', 'D', 'A', 'X'], dtype=object)

Agora vamos verificar as cores das provas

In [48]:
itens_provas_mt_2010_2019['TX_COR'].value_counts().sort_index()

AMARELA          360
AMARELO           45
AZUL             405
Amarela          180
Amarelo           90
Azul             225
Azul              45
CINZA            495
Cinza            225
Cinza             90
Cinza (Ledor)     45
LARANJA           45
Laranja           90
ROSA             405
Rosa             225
Rosa              45
VERDE             45
Verde             90
Name: TX_COR, dtype: int64

Para padronizar os nomes das cores, vamos retirar o texto "(Ledor)" da cor "Cinza (Ledor)"

In [49]:
itens_provas_mt_2010_2019['TX_COR'][itens_provas_mt_2010_2019['TX_COR'] == 'Cinza (Ledor)'] = 'Cinza'

Agora vamos colocar todos os nomes com letras maíusculas

In [50]:
itens_provas_mt_2010_2019['TX_COR'] = itens_provas_mt_2010_2019['TX_COR'].str.upper()

Mostrando as cores novamente

In [51]:
itens_provas_mt_2010_2019['TX_COR'].value_counts().sort_index()

AMARELA    540
AMARELO    135
AZUL       630
AZUL        45
CINZA      765
CINZA       90
LARANJA    135
ROSA       630
ROSA        45
VERDE      135
Name: TX_COR, dtype: int64

Agora vamos mudar a cor "AMARELO" para "AMARELA"

In [52]:
itens_provas_mt_2010_2019['TX_COR'][itens_provas_mt_2010_2019['TX_COR'] == 'AMARELO'] = 'AMARELA'

Agora vamos verificar a razão das cores "AZUL", "CINZA" e "ROSA" aparecerem duas vezes

In [53]:
itens_provas_mt_2010_2019['TX_COR'].unique()

array(['AMARELA', 'CINZA', 'AZUL', 'ROSA', 'CINZA  ', 'ROSA   ',
       'AZUL   ', 'LARANJA', 'VERDE'], dtype=object)

Como pode ser observado, as cores aparecem duas vezes porque estão com espaços em branco no final. Vamos corrigir.

In [54]:
itens_provas_mt_2010_2019['TX_COR'] = itens_provas_mt_2010_2019['TX_COR'].str.strip()

Mostrando novamente as cores após os ajustes

In [55]:
itens_provas_mt_2010_2019['TX_COR'].value_counts().sort_index()

AMARELA    675
AZUL       675
CINZA      855
LARANJA    135
ROSA       675
VERDE      135
Name: TX_COR, dtype: int64

Verificando se os valores dos códigos de habilidade são números de 1 a 30

In [56]:
itens_provas_mt_2010_2019['CO_HABILIDADE'].unique()

array([1, 7, 11, 12, 25, 15, 26, 24, 10, 6, 20, 19, 17, 9, 16, 4, 3, 30,
       14, 23, 8, 22, 21, 27, 13, 5, 29, 28, 18, 2, ' 1', '10', '21',
       ' 5', '17', '12', '13', '11', ' 7', '27', '22', '23', ' 8', '18',
       '15', ' 4', '25', '19', ' 2', ' 3', '30', '16', '26', ' 9', '29',
       '28', '24', ' 6', '20', '14'], dtype=object)

Modificando o tipo dos dados da coluna "CO_HABILIDADE"

In [57]:
itens_provas_mt_2010_2019['CO_HABILIDADE'] = itens_provas_mt_2010_2019['CO_HABILIDADE'].astype('int64')

Vamos verificar se a alteração foi realizada

In [58]:
itens_provas_mt_2010_2019['CO_HABILIDADE'].unique()

array([ 1,  7, 11, 12, 25, 15, 26, 24, 10,  6, 20, 19, 17,  9, 16,  4,  3,
       30, 14, 23,  8, 22, 21, 27, 13,  5, 29, 28, 18,  2], dtype=int64)

Agora vamos verificar os códigos das provas 

In [59]:
itens_provas_mt_2010_2019['CO_PROVA'].unique()

array([97, 98, 99, 100, '130', '129', '132', '131', '136', 149, 150, 151,
       152, 156, 179, 180, 181, 182, 207, 208, 209, 210, 243, 244, 245,
       246, 304, 306, 303, 305, 310, 347, 349, 346, 348, 367, 369, 366,
       368, 370, 404, 406, 403, 405, 410, 443, 444, 445, 446, 414, 461,
       459, 462, 460, 470, 466, 499, 500, 501, 502, 515, 516, 517, 518,
       522, 526, 555, 556, 557, 558], dtype=object)

Os códigos das provas 129 a 136 foram armazenados na forma de texto (string). Desta forma, vamos mudar os tipos dos valores para numeros inteiros

In [60]:
itens_provas_mt_2010_2019['CO_PROVA'] = itens_provas_mt_2010_2019['CO_PROVA'].astype('int64')

Vamos verificar se a alteração foi realizada

In [61]:
itens_provas_mt_2010_2019['CO_PROVA'].unique()

array([ 97,  98,  99, 100, 130, 129, 132, 131, 136, 149, 150, 151, 152,
       156, 179, 180, 181, 182, 207, 208, 209, 210, 243, 244, 245, 246,
       304, 306, 303, 305, 310, 347, 349, 346, 348, 367, 369, 366, 368,
       370, 404, 406, 403, 405, 410, 443, 444, 445, 446, 414, 461, 459,
       462, 460, 470, 466, 499, 500, 501, 502, 515, 516, 517, 518, 522,
       526, 555, 556, 557, 558], dtype=int64)